src/forest.rs

```diff
@@ -8,25 +8,63 @@ use std::hash::Hash;
 use std::io::{self, Write};
 use std::str;
 
+/// Some sort of object capable of providing information about various parts of
+/// the grammar (mostly parse nodes).
+/// For code generation, this doesn't need to be more than an unit struct, as
+/// all the information can be hardcoded, but in more dynamic settings, this
+/// might contain e.g. a reference to a context.
+pub trait GrammarReflector {
+    type ParseNodeKind: fmt::Debug + Ord + Hash + Copy;
+
+    fn parse_node_shape(&self, kind: Self::ParseNodeKind) -> ParseNodeShape<Self::ParseNodeKind>;
+    fn parse_node_desc(&self, kind: Self::ParseNodeKind) -> String;
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ParseNode<'i, P> {
+    pub kind: P,
+    pub range: Range<'i>,
+}
+
+impl<P: fmt::Debug> fmt::Debug for ParseNode<'_, P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{:?} @ {}..{}",
+            self.kind,
+            self.range.start(),
+            self.range.end()
+        )
+    }
+}
+
 /// A parse forest, in SPPF (Shared Packed Parse Forest) representation.
-pub struct ParseForest<'i, P: ParseNodeKind, I: Input> {
+pub struct ParseForest<'i, G: GrammarReflector, I: Input> {
+    pub grammar: G,
     // HACK(eddyb) `pub(crate)` only for `parser`.
     pub(crate) input: Container<'i, I::Container>,
-    pub(crate) possible_choices: HashMap<ParseNode<'i, P>, BTreeSet<P>>,
-    pub(crate) possible_splits: HashMap<ParseNode<'i, P>, BTreeSet<usize>>,
+    pub(crate) possible_choices:
+        HashMap<ParseNode<'i, G::ParseNodeKind>, BTreeSet<G::ParseNodeKind>>,
+    pub(crate) possible_splits: HashMap<ParseNode<'i, G::ParseNodeKind>, BTreeSet<usize>>,
```

Comment:
These two really want a type alias =P The only diff is `usize` vs. `G::ParseNodeKind`.

Reply:
I'd rather not, unless a common API can be built.
I've been wanting to optimize this for the one-element case, or use a bitset, for a while now, but I don't want to do it right now.